runsc/fsgofer/fsgofer.go

```diff
@@ -124,30 +127,47 @@ func (a *attachPoint) Attach() (p9.File, error) {
 	if err != nil {
 		return nil, fmt.Errorf("stat file %q, err: %v", a.prefix, err)
 	}
-	mode := syscall.O_RDWR
-	if a.conf.ROMount || (stat.Mode&syscall.S_IFMT) == syscall.S_IFDIR {
-		mode = syscall.O_RDONLY
-	}
 
-	// Open the root directory.
-	f, err := fd.Open(a.prefix, openFlags|mode, 0)
-	if err != nil {
-		return nil, fmt.Errorf("unable to open file %q, err: %v", a.prefix, err)
+	// Hold the file descriptor we are converting into a p9.File
+	var f *fd.FD
+
+	// Apply the S_IFMT bitmask so we can detect file type appropriately
+	switch fmtStat := stat.Mode & syscall.S_IFMT; {
+	case fmtStat == syscall.S_IFSOCK:
```

Comment:
nit:
case syscall.S_IFSOCK:

Reply:
This is actually the proper case, and the suggested change would be invalid due to grammatical semantics.
```go
switch fmtStat := stat.Mode & syscall.S_IFMT; {...
```
evaluates to the equivalent of
```go
switch ; {...
```
Or, more succinctly:
```go
switch {...
```
This results in an implicit `true` being compared to the result of each `case` expression, semantically equivalent to:
```go
case (true) == (fmtStat == syscall.S_IFSOCK): ...
```
If we were to perform `case syscall.S_IFSOCK` there would be a type error when attempting to compare a `bool` and `int`, because it would be semantically equivalent to:
```go
case (true) == (syscall.S_IFSOCK): ...
```

Reply:
You need to change the switch header too:
```
switch stat.Mode & syscall.S_IFMT {
```

Reply:
By doing that it would require expanding:
https://github.com/google/gvisor/blob/c2ae77a607b6e103545aa83e8f2c7c5bf649846f/runsc/fsgofer/fsgofer.go#L149-L151
Which is why it was put in the switch statement.

Reply:
You can do both:
```go
switch fmtStat := stat.Mode & syscall.S_IFMT; fmtStat {...
```

Reply:
Done as @iangudger suggested.